### PR TITLE
Fix memory leak.

### LINKED ISSUE #3505: Memory leak in Google PubSub python client
I am publishing messages indefinitely to Google PubSub using Python. I am publishing the messages in batches. I've set the batch_size as 20. I am publishing messages in 5 processes.

I executed below script in GCE with 1 CPU and 3.75 memory. After running for more than a hour, the GCE is out of memory and the publisher get crashed.

```
import json

from multiprocessing import Process
from google.cloud import pubsub

_PROJECT = 'google-project-id'
_TOPIC = 'sample'
_PROCESS_COUNT = 5


def publish(message_str, ps_batch):
    while True:
        ps_batch.publish(message_str)

def main():
    json_message = {'id': 1,'first_name': 'Karim','last_name': 'Bourdice','email': 'kbourdice0@purevolume.com','gender': 'Male','ip_address': '76.108.207.137','field1': '1MCwggd9g267ZBfq8p6KUR8RzXh1UARcBM','field2': '14MbkGizhAuYY9fjzA7UQ1x5sHfPEearMy','field3': '1BC7imHVSwT2xKmZBE6xVwGExXte4M87L7','field4': '1H2CPcuv9JvmnKV3Q9zxtZqZkvn86yyqFz','field5': '1GkanWfbWXxmriDt4vdDq5pb2wySL1nDiZ','field6': '1VKh8m3VcEt5xkfqj9BcCJfuL5nVc3nKZ','field7': '1NuruyCxMeL6YM8ouzD9aqAqFDfHzb6pak','field8': '1DB3Ac7uR5NJziSfJVLM1G2F6tW27SkZqN','field9': '1LdM4nGXPQkSKPTLkZFTbKHN9sAUpBGtSf','field10': '1LLCjdpVYy97u7kBdhHJcgycAZbdZ3R1xF','field11': '16FaG4qFomWMmv4Zv4F8MMUBkcJcEHHijY','field12': '1PXkwtCc4CWSp8BkecAvQubMs12GCvqdVu','field13': '1BqnENPGGDXn9S1FHn2pvwLtDDYu1QTjHL','field14': '1KckpoksBagHxANikRrRBC428srpxAqY5C','field15': '17Z3yPdFgkxEn29JaT89HshvVF4RbtMsMq','field16': '14QpaoakehbRnckiQd7Zte2nSRRGX695Wk','field17': '1wGqgYDSmfE9Q1ZH67GwLqNQmH2u7viRs','field18': '134KjLvr8ugwZK19wzL3V9i8fvmaBBjHQM','field19': '1JrUeWAvufRHXbMLSQeGLkjHJNx3XFM77A','field20': '19Botyc2ikY7V9XqRkJbnoJHWk2eVRxni5','field29': '1DjHXt1yecWxjuNPDMQDaAiCjBvDoEwZr1','field21': '1AZejYQWuuQi9DsGp27tHt7PdCRyBtLD8L','field22': '1AzK3J6vwwh8Xog5TJA48Uh223m5LwqRxz','field23': '12rSBAhFvgFMWbssYmSF3zCpDxmpSPxJV6','field24': '1C1QxVG8QphiRD2uzF6Mg86vRdwTARXpeW','field25': '1GRCqpUZ3oejbic59z16XWUYbw8GYdiq6f','field26': '1CH2yzHjehfuyVW9vw9NPTWX3yW5v2vTFg','field27': '1CCoa5aQPJ4Ya1FDKWrMceRReyptKcWV1N','field28': '1LJSmvzGWwDBLc1YTPaT2k1uAkiX8GjGpz'}
    message_str = json.dumps(json_message)
    batch_size = 20

    ps = pubsub.Client()
    ps_topic = ps.topic(_TOPIC)
    ps_batch = ps_topic.batch(max_messages = batch_size)
    print("Pushing messages to {0}".format(_TOPIC))

    processes = []
    for i in range(1, _PROCESS_COUNT):
        p = Process(target=publish, args=(message_str, ps_batch))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()


if __name__ == "__main__":
    main()
```

Here is the stackdriver gce-memory-used chart. enter image description here
stackdriver-publisher-crash

You can see that memory-used is steadily increasing after running the above python script. And finally it got crashed. Please note that I am running this python script alone in this GCE and not any other application

I've asked this in [Stackoverflow](https://stackoverflow.com/questions/44542739/memory-leak-in-google-pubsub-python-client) as well. Answer from Luke claims the message_ids is not cleaned up over [here](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/pubsub/google/cloud/pubsub/topic.py#L550). I feel this is an issue and hence I created this issue in addition to stackoverflow.

Please note that this issue is migrated from [here](https://issuetracker.google.com/issues/62606697)

### OS Details  
```Ubuntu 16.10```
### Python Version
```
$ python --version
Python 2.7.12+
```
### google-could-python version
```
$ pip freeze
boto==2.40.0
cachetools==2.0.0
certifi==2017.4.17
chardet==3.0.4
crcmod==1.7
dill==0.2.6
enum34==1.1.6
future==0.16.0
futures==3.1.1
gapic-google-cloud-pubsub-v1==0.15.4
google-auth==1.0.1
google-auth-httplib2==0.0.2
google-cloud-bigquery==0.24.0
google-cloud-core==0.24.1
google-cloud-pubsub==0.25.0
google-compute-engine==2.2.4
google-gax==0.15.13
googleapis-common-protos==1.5.2
grpc-google-iam-v1==0.11.1
grpcio==1.3.5
httplib2==0.10.3
idna==2.5
memory-profiler==0.47
netifaces==0.10.4
oauth2client==3.0.0
ply==3.8
proto-google-cloud-pubsub-v1==0.15.4
protobuf==3.3.0
psutil==4.2.0
pyasn1==0.2.3
pyasn1-modules==0.0.9
requests==2.17.3
rsa==3.4.2
six==1.10.0
urllib3==1.21.1

```

### PR DESCRIPTION
**Pull request by lukesneeringer**
This changes the way message_ids are returned. Instead of being populated into the batch, they are returned sometimes (if a commit was triggered) in each publish.

Fixes #3505

This addresses the memory leak surfaced in: https://stackoverflow.com/questions/44542739/

This represents a backwards incompatible change, but I expect nobody was trying to read the `message_ids` since they were surfaced so poorly, and furthermore I am working on a client that supercedes this anyway.

### CODE REVIEW

File: pubsub/google/cloud/pubsub/topic.py [hunk 1]
```diff
@@ -499,6 +498,10 @@ def publish(self, message, **attrs):
 
         :type attrs: dict (string -> string)
         :param attrs: key-value pairs to send as message attributes
+
+        :rtype: list
+        :returns: A list, potentially with message IDs (if the batch
+            committed).
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub/topic.py]: Typically we've aligned the content

**Comment by lukesneeringer** [comment]
The previous version stacked up a list with an ever-increasing list of strings on the class, so if you run a batch forever, the list keeps growing until you OOM.

This version returns the list, and if you do not actively store it yourself, then the garbage collector will eat it.

File: pubsub/google/cloud/pubsub/topic.py [hunk 2]
```diff
@@ -547,5 +550,5 @@ def commit(self, client=None):
             client = self.client
         api = client.publisher_api
         message_ids = api.topic_publish(self.topic.full_name, self.messages[:])
-        self.message_ids.extend(message_ids)
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub/topic.py]: @lukesneeringer This is the line I failed to see

**Comment by tseaver** [comment]
Note also that "run a batch forever" was not an intended usecase: PR #3157 (which added the `max_size` auto-commit) introduced that as a theoretical possibility, but it was never documented to be used so.

**Comment by dhermes** [comment]
@lukesneeringer Shouldn't this be merged?

**Comment by tseaver** [comment]
@dhermes it still lacks doc updates and test changes.

**Comment by lukesneeringer** [comment]
I am also not worrying about getting it in since we are trying to land an entirely new library this week that wipes it out anyway. Probably easier to close this than land it.

**Pull request closed without merging**